src/awkward/_connect/numba/builder.py

```diff
@@ -396,6 +463,105 @@ def lower_real(context, builder, sig, args):
+        z_real = builder.sitofp(xval, context.get_value_type(numba.types.float64))
+        z_imag = z_real.type(0)
+    elif isinstance(xtype, numba.types.Integer):
+        z_real = builder.uitofp(xval, context.get_value_type(numba.types.float64))
+        z_imag = z_real.type(0)
+    elif xtype.bitwidth < 64:
+        z_real = builder.fpext(xval, context.get_value_type(numba.types.float64))
+    elif xtype.bitwidth > 64:
+        z_real = builder.fptrunc(xval, context.get_value_type(numba.types.float64))
+    else:
+        z_real = xval
+    z_imag = z_real.type(0)
+
+    call(
+        context,
+        builder,
+        ak._libawkward.ArrayBuilder_complex,
+        (proxyin.rawptr, z_real, z_imag),
+    )
+    return context.get_dummy_value()
+
+
+@numba.extending.lower_builtin("complex", ArrayBuilderType, numba.types.Complex)
+def lower_complex(context, builder, sig, args):
+    arraybuildertype, xtype = sig.args
+    arraybuilderval, xval = args
+    proxyin = context.make_helper(builder, arraybuildertype, arraybuilderval)
+
+    z = context.make_complex(builder, xtype, xval)
+    z_real, z_imag = z.real, z.imag
+
+    call(
+        context,
+        builder,
+        ak._libawkward.ArrayBuilder_complex,
+        (proxyin.rawptr, z_real, z_imag),
+    )
+    return context.get_dummy_value()
+
+
+@numba.extending.lower_builtin("datetime", ArrayBuilderType, numba.types.NPDatetime)
+def lower_datetime(context, builder, sig, args):
+    arraybuildertype, xtype = sig.args
+    arraybuilderval, xval = args
+    proxyin = context.make_helper(builder, arraybuildertype, arraybuilderval)
+    unit = globalstring(context, builder, f"datetime64[{xtype.unit}]")
+    call(
+        context,
+        builder,
+        ak._libawkward.ArrayBuilder_datetime,
+        (proxyin.rawptr, xval, unit),
+    )
+    return context.get_dummy_value()
+
+
+@numba.extending.lower_builtin("timedelta", ArrayBuilderType, numba.types.NPTimedelta)
+def lower_timedelta(context, builder, sig, args):
+    arraybuildertype, xtype = sig.args
+    arraybuilderval, xval = args
+    proxyin = context.make_helper(builder, arraybuildertype, arraybuilderval)
+    unit = globalstring(context, builder, f"timedelta64[{xtype.unit}]")
+    call(
+        context,
+        builder,
+        ak._libawkward.ArrayBuilder_timedelta,
+        (proxyin.rawptr, xval, unit),
+    )
+    return context.get_dummy_value()
+
+
+@numba.extending.lower_builtin("string", ArrayBuilderType, numba.types.UnicodeType)
+def lower_string(context, builder, sig, args):
+    arraybuildertype, xtype = sig.args
+    arraybuilderval, xval = args
+    proxyin = context.make_helper(builder, arraybuildertype, arraybuilderval)
+
+    pyapi = context.get_python_api(builder)
+    gil = pyapi.gil_ensure()
+
+    out = pyapi.string_as_string(xval.value)
+    call(
+        context,
+        builder,
+        ak._libawkward.ArrayBuilder_string,
+        (proxyin.rawptr, out),
+    )
```

Comment:
Even though the Python API function ensures that the `char*` string ends with a null terminator, the string might still have null bytes within the string. For instance,

```python
>>> b"arrow \xe2\x86\x92 zero \x00 not the end!".decode("utf-8")
'arrow → zero \x00 not the end!'
>>> b"arrow \xe2\x86\x92 zero \x00 not the end!".decode("utf-8").encode("utf-8")
b'arrow \xe2\x86\x92 zero \x00 not the end!'
```

Unless we explicitly ask for a length, we wouldn't be able to tell that the above string isn't `"arrow → zero "`. The ArrayBuilder has a assume-null-terminator method for convenience, but also a `char*`-and-length method for correctness:

https://github.com/scikit-hep/awkward/blob/38243b2f75fbf559bf31f97dfaeb02c428645c59/include/awkward/builder/ArrayBuilder.h#L304-L313

To use it, we also have to ask Python for the length. ([PythonAPI.string_as_string_and_size](https://github.com/numba/numba/blob/f4b6fe0124fa7734ba2bb28c7685db5b6306c486/numba/core/pythonapi.py#L1079-L1103), which is [PyUnicode_AsUTF8AndSize](https://docs.python.org/3/c-api/unicode.html#c.PyUnicode_AsUTF8AndSize), and that size is the number of _bytes_, not the number of code-points, so that's good.)

```suggestion
    is_ok, out, length = pyapi.string_as_string_and_size(xval.value)
    length = ak._connect.numba.layout.castint(context, builder, numba.ssize_t, numba.int64, length)
    call(
        context,
        builder,
        ak._libawkward.ArrayBuilder_string_length,
        (proxyin.rawptr, out, length),
    )

```